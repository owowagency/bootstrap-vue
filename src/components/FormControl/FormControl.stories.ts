import FormControl from '.';
import {ref} from 'vue';
import {sizes} from '@/composables/useSize';
import {types} from '@/composables/useFormControl';

export default {
    title: 'Forms/Form control',
    component: FormControl,
    argTypes: {
        disabled: {
            control: {type: 'boolean'},
        },
        placeholder: {
            control: {type: 'text'},
        },
        plainText: {
            control: {type: 'boolean'},
        },
        readonly: {
            control: {type: 'boolean'},
        },
        size: {
            control: {type: 'select'},
            options: sizes,
        },
        type: {
            control: {type: 'select'},
            options: types,
        },
    },
};

const template = (args) => ({
    components: {FormControl},
    setup() {
        return {args, value: ref('')};
    },
    template: '<FormControl v-model="value" v-bind="args" />',
});

export const Default = template.bind({});

Default.args = {
    modelValue: 'I am an input',
    placeholder: 'I am a placeholder',
};

export const Autofocused = template.bind({});

Autofocused.args = {
    modelValue: 'I am autofocused',
    autofocus: true,
};

export const Disabled = template.bind({});

Disabled.args = {
    modelValue: 'I am disabled',
    disabled: true,
};

export const PlainText = template.bind({});

PlainText.args = {
    modelValue: 'Plain text',
    plainText: true,
};

export const Readonly = template.bind({});

Readonly.args = {
    modelValue: 'I am readonly',
    readonly: true,
};

export const Type = (args) => ({
    components: {FormControl},
    setup() {
        return {args, types};
    },
    template: '<FormControl v-for="t of types" class="mb-1" :placeholder="t" :type="t" v-bind="args" />',
});
