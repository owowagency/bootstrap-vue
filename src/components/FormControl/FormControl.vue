<template>
    <input
        ref="input"
        :autofocus="autofocus"
        class="form-control"
        :class="classes"
        :disabled="disabled"
        :placeholder="placeholder"
        :readonly="readonly"
        :required="required"
        :type="type"
        :value="modelValue"
        @input="emit('update:modelValue', $event.target.value)"
    >
</template>

<script lang="ts">
import {computed, ref} from 'vue';
import {formControlProps} from '@/composables/useFormControl';
import useClasses from '@/composables/useClasses';
import useDisabled from '@/composables/useDisabled';
import useSize from '@/composables/useSize';
</script>

<script lang="ts" setup>
const props = defineProps(formControlProps);

const emit = defineEmits([
    /**
     * Fired when the model value is updated
     * @param {any} value The new value
     */
    'update:modelValue',
]);

const input = ref<HTMLInputElement>();

const {classes} = useClasses(computed(() => [
    useDisabled(props.disabled).disabledClass.value,
    props.plainText ? 'form-control-plaintext' : '',
    props.readonly ? 'readonly' : '',
    useSize(props.size, 'form-control-{0}').sizeClass.value,
]));

defineExpose({input});
</script>

<docs>
```vue
<FormControl
    name="username"
    v-model="username"
    placeholder="JohnDoe"
/>
```

## Imported Props

| Prop name   | Description                                  | Type                                              | Values | Default |
| ----------- | -------------------------------------------- | ------------------------------------------------- | ------ | ------- |
| autofocus   | Automatically focus on the input when loaded | boolean                                           | -      | false   |
| disabled    | Disables the component                       | boolean                                           | -      | false   |
| modelValue  | The model value                              | string                                            | -      | ''      |
| placeholder | The placeholder of the form input            | string                                            | -      | ''      |
| plainText   | Adds the `form-control-plaintext` class      | boolean                                           | -      | false   |
| readonly    | Makes the component readonly                 | boolean                                           | -      | false   |
| size        | The size of the component                    | [Size](../../composables/useSize) (string)        | -      | 'md'    |
| type        | The type of the component                    | [Type](../../composables/useFormControl) (string) | -      | 'text'  |
</docs>
