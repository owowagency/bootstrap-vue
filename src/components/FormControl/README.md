# FormControl

```vue
<FormControl name="username" v-model="username" placeholder="JohnDoe" />
```

## Imported Props

| Prop name   | Description                                  | Type                                              | Values | Default |
| ----------- | -------------------------------------------- | ------------------------------------------------- | ------ | ------- |
| autofocus   | Automatically focus on the input when loaded | boolean                                           | -      | false   |
| disabled    | Disables the component                       | boolean                                           | -      | false   |
| modelValue  | The model value                              | string                                            | -      | ''      |
| placeholder | The placeholder of the form input            | string                                            | -      | ''      |
| plainText   | Adds the `form-control-plaintext` class      | boolean                                           | -      | false   |
| readonly    | Makes the component readonly                 | boolean                                           | -      | false   |
| size        | The size of the component                    | [Size](../../composables/useSize) (string)        | -      | 'md'    |
| type        | The type of the component                    | [Type](../../composables/useFormControl) (string) | -      | 'text'  |

## Events

| Event name        | Properties                      | Description                           |
| ----------------- | ------------------------------- | ------------------------------------- |
| update:modelValue | **value** `any` - The new value | Fired when the model value is updated |
