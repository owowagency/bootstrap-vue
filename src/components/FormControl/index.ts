import FormControl from './FormControl.vue';

export default FormControl;
