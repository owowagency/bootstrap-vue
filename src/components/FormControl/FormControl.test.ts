import FormControl from '.';
import {shallowMount} from '@vue/test-utils';

describe('template', () => {
    componentRenderTest(FormControl);

    it('adds autofocus property', () => {
        const wrapper = shallowMount(FormControl, {
            props: {autofocus: true},
        });

        expect((wrapper.element as HTMLInputElement).autofocus).toBe(true);
    });

    it('adds disabled property', () => {
        const wrapper = shallowMount(FormControl, {
            props: {disabled: true},
        });

        expect((wrapper.element as HTMLInputElement).disabled).toBe(true);
    });

    it('adds placeholder property', () => {
        const wrapper = shallowMount(FormControl, {
            props: {placeholder: 'I am a placeholder'},
        });

        expect(wrapper.attributes('placeholder')).toBe('I am a placeholder');
    });

    it('adds readonly property', () => {
        const wrapper = shallowMount(FormControl, {
            props: {readonly: true},
        });

        expect((wrapper.element as HTMLInputElement).readOnly).toBe(true);
    });

    componentWrapperClassTest(FormControl, {disabled: true}, 'disabled');

    componentWrapperClassTest(FormControl, {plainText: true}, 'form-control-plaintext');

    componentWrapperClassTest(FormControl, {readonly: true}, 'readonly');

    componentWrapperClassTest(FormControl, {size: 'lg'}, 'form-control-lg');
});
