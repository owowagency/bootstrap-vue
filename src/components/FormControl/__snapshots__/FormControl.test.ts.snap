// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`template > renders default 1`] = `"<input class=\\"form-control form-control-md\\" placeholder=\\"\\" type=\\"text\\">"`;
